### DISCUSSION TITLE
[Ideas] [2025-01-28] BPR Loss + Graph Regularization Loss

### DISCUSSION THREAD

**Discussion by annakong23** [boostcampaitech7/level4-recsys-finalproject-hackathon-recsys-05-lv3#32] [Idea]
### 아이디어 제목

BPR Loss에 Graph Regularization Loss 추가

### '아이템 임베딩이 너무 멀어지지 않도록 정규화하는 역할' 을 하는 Graph Regularization Loss 추가
- BPR Loss (self.model.bpr_loss(users, pos, neg))→ 기존 BPR Loss 계산
- Graph Regularization Loss (graph_regularization_loss())→ 아이템 간 유사도가 높은 경우 임베딩이 가까워지도록 정규화
- Similarity Matrix (compute_similarity_matrix())→ Cosine Similarity를 사용해 아이템 임베딩 간 유사도 계산
- 최종 Loss (total_loss = loss + reg_loss + graph_reg_loss)→ 기존 Loss + 정규화 Loss 적용 후 최적화

### BPRLossWithReg 클래스
```
class BPRLossWithReg:
    def __init__(self,
                 recmodel : PairWiseModel,
                 config : dict):
        self.model = recmodel
        self.weight_decay = config['decay']
        self.reg_weight = config.get('reg_weight', 0.01)  # 정규화 가중치 추가
        self.lr = config['lr']
        self.opt = optim.Adam(recmodel.parameters(), lr=self.lr)

    def stageOne(self, users, pos, neg):
        """
        BPR Loss + Regularization Loss 적용.
        """
        # 기존 BPR Loss 계산
        loss, reg_loss = self.model.bpr_loss(users, pos, neg)
        reg_loss = reg_loss * self.weight_decay
        
        # 추가: Graph Regularization Loss 계산
        graph_reg_loss = self.graph_regularization_loss()
        graph_reg_loss = graph_reg_loss * self.reg_weight  # 가중치 적용

        # 최종 Loss 계산
        total_loss = loss + reg_loss + graph_reg_loss

        # 역전파 및 최적화
        self.opt.zero_grad()
        total_loss.backward()
        self.opt.step()

        return total_loss.cpu().item()

    def graph_regularization_loss(self):
        """
        Graph Regularization Loss 계산.
        L_reg = Σ w_ij ||E_i - E_j||^2
        """
        item_embeddings = self.model.item_embedding.weight  # 아이템 임베딩 가져오기
        similarity_matrix = self.compute_similarity_matrix(item_embeddings)  
        reg_loss = torch.sum(similarity_matrix * (item_embeddings.unsqueeze(1) - item_embeddings.unsqueeze(0)).pow(2))

        return reg_loss

    def compute_similarity_matrix(self, embeddings):
        """
        Cosine Similarity 기반으로 아이템 간 유사도 행렬 계산.
        """
        norm_embeddings = F.normalize(embeddings, p=2, dim=1)  # 정규화
        similarity_matrix = torch.mm(norm_embeddings, norm_embeddings.t())  # Cosine Similarity 계산

        return similarity_matrix
```
### BPR Loss 부분에서 기존과 다른 점

1. self.reg_weight의 역할  
- `self.reg_weight`는 Graph Regularization Loss에 대한 가중치 역할.  
- 만약 이 가중치가 없으면, 정규화 손실값이 너무 커져서 BPR Loss가 학습을 방해할 가능성이 있음.  
- 너무 작으면 Graph Regularization Loss의 효과가 없어짐.  
  
2. self.reg_weight를 유지하는 이유  
- LightGCN에서 BPR Loss가 기본적으로 사용되므로, Graph Regularization Loss가 너무 강하면 학습이 왜곡될 가능성이 있음.  
- 이를 방지하기 위해 self.reg_weight를 사용해서 적절한 비율로 정규화 Loss를 반영.  

### 예상 효과

Graph Regularization Loss는 아이템 임베딩이 너무 멀어지지 않도록 함
=유사한 관계(예: 같은 카테고리, 유사한 설명을 가진 아이템)의 아이템들이 서로 가까운 임베딩을 갖도록 만드는 역할

**필요한 이유**
- 새로운 아이템(Cold-start 아이템)은 사용자와의 상호작용이 부족해서, 모델이 적절한 임베딩을 찾기 어려울 수 있음
- 이때 정규화를 통해 기존 아이템과 비슷한 임베딩을 가지도록 강제하면, 새로운 아이템이 더 좋은 추천 성능을 가질 가능성이 높아짐

### 생각해볼 내용
이 loss fuction을 사용하는 경우 아이템 메타데이터 임베딩을 어느정도까지 해야하는지, 아니면 아예 하지 않아도 되는 것인지 생각해보기

### 참고 자료

눈문[Neural Graph Collaborative Filtering]
https://arxiv.org/pdf/1905.08108
